### PR TITLE
nixos/cloud-init: add support for network configuration

### PR DESCRIPTION
**Pull request by illustris**
###### Motivation for this change
Cloud-init on nixos does not support network interface configuration. This PR adds that functionality.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by illustris** [comment]
```
[nix-shell:~/src/nixpkgs]$ cat ./testvm.nix
{ ... }:
{
        config = {
                services = {
                        cloud-init = {
                                enable = true;
                                network.enable = true;
                        };
                        openssh.enable = true;
                        getty.autologinUser = "root";
                };
                security.sudo.wheelNeedsPassword = false;
                networking.useDHCP = false;
                users.users.illustris = {
                        isNormalUser = true;
                        extraGroups = [ "wheel" ];
                        openssh.authorizedKeys.keyFiles = [ /home/illustris/.ssh/id_rsa.pub ];
                };
        };
}

[nix-shell:~/src/nixpkgs]$ nixos-generate -I nixpkgs=$(pwd) --format proxmox --configuration ./testvm.nix
.
.
.
[    2.801365] reboot: Power down
vzdump-qemu-nixos-22.05.git.d41882c7b98M.vma : 27.28%   (2026543104 => 552818826 bytes, vzdump-qemu-nixos-22.05.git.d41882c7b98M.vma.zst)
/nix/store/6ps99m2g2dqslbyv7rag5imvwyc7lcba-proxmox-nixos-22.05.git.d41882c7b98M/vzdump-qemu-nixos-22.05.git.d41882c7b98M.vma.zst

[nix-shell:~/src/nixpkgs]$ rsync -avzh --progress /nix/store/6ps99m2g2dqslbyv7rag5imvwyc7lcba-proxmox-nixos-22.05.git.d41882c7b98M/vzdump-qemu-nixos-22.05.git.d41882c7b98M.vma.zst root@cube:/var/lib/vz/dump/
.
.
sent 490.51M bytes  received 164.72K bytes  8.68M bytes/sec
total size is 552.82M  speedup is 1.13

[nix-shell:~/src/nixpkgs]$ ssh root@cube

root@cube:~# qmrestore /var/lib/vz/dump/vzdump-qemu-nixos-22.05.git.d41882c7b98M.vma. 500
restore vma archive: zstd -q -d -c /var/lib/vz/dump/vzdump-qemu-nixos-22.05.git.d41882c7b98M.vma.zst | vma extract -v -r /var/tmp/vzdumptmp28315.fifo - /var/tmp/vzdumptmp28315
CFG: size: 285 name: qemu-server.conf
DEV: dev_id=1 size: 3629121536 devname: drive-virtio0
CTIME: Sun Jan 30 09:01:58 2022
  Rounding up size to full physical extent 3.38 GiB
  Logical volume "vm-500-disk-0" created.
new volume ID is 'local-lvm:vm-500-disk-0'
map 'drive-virtio0' to '/dev/pve/vm-500-disk-0' (write zeros = 0)
progress 1% (read 36306944 bytes, duration 0 sec)
.
.
.
rescan volumes...
VM 500 (virtio0): size of disk 'local-lvm:vm-500-disk-0' updated from 0T to 3464M

root@cube:~# qm set 500 --ide2 local-lvm:cloudinit
update VM 500: -ide2 local-lvm:cloudinit
  Logical volume "vm-500-cloudinit" created.

root@cube:~# qm set 500 --ipconfig0 ip=192.168.33.11/24,gw=192.168.33.1
update VM 500: -ipconfig0 ip=192.168.33.11/24,gw=192.168.33.1

root@cube:~# cat /etc/pve/qemu-server/500.conf
# generated by NixOS
agent: 1
boot:
cores: 1
ide2: local-lvm:vm-500-cloudinit,media=cdrom
ipconfig0: ip=192.168.33.11/24,gw=192.168.33.1
memory: 1024
name: nixos-22.05.git.d41882c7b98M
net0: virtio=12:5A:CE:B3:73:C9,bridge=vmbr0,firewall=1
ostype: l26
scsihw: virtio-scsi-pci
serial0: socket
virtio0: local-lvm:vm-500-disk-0,size=3464M

root@cube:~# qm cloudinit dump 500 network
version: 1
config:
    - type: physical
      name: eth0
      mac_address: '12:5a:ce:b3:73:c9'
      subnets:
      - type: static
        address: '192.168.33.11'
        netmask: '255.255.255.0'
        gateway: '192.168.33.1'
    - type: nameserver
      address:
      - '8.8.8.8'
      search:
      - 'example.com'

root@cube:~# qm start 500; qm terminal 500
generating cloud-init ISO
starting serial terminal on interface serial0 (press Ctrl+O to exit)

<<< NixOS Stage 1 >>>

loading module dm_mod...
running udev...
.
.
.
[root@nixos:~]# ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
    link/ether 12:5a:ce:b3:73:c9 brd ff:ff:ff:ff:ff:ff
    altname enp0s18
    altname ens18
    inet 192.168.33.11/24 brd 192.168.33.255 scope global eth0
       valid_lft forever preferred_lft forever

[root@nixos:~]# ping 8.8.8.8
PING 8.8.8.8 (8.8.8.8) 56(84) bytes of data.
64 bytes from 8.8.8.8: icmp_seq=1 ttl=53 time=25.3 ms
64 bytes from 8.8.8.8: icmp_seq=2 ttl=53 time=24.9 ms
64 bytes from 8.8.8.8: icmp_seq=3 ttl=53 time=24.0 ms
^C
--- 8.8.8.8 ping statistics ---
4 packets transmitted, 3 received, 25% packet loss, time 3005ms
rtt min/avg/max/mdev = 24.001/24.754/25.337/0.558 ms
```

File: pkgs/tools/virtualization/cloud-init/default.nix [hunk 1]
```diff
@@ -25,6 +28,13 @@ python3.pkgs.buildPythonApplication rec {
   prePatch = ''
     substituteInPlace setup.py \
       --replace /lib/systemd $out/lib/systemd
+
+    substituteInPlace cloudinit/net/networkd.py \
+      --replace "['/usr/sbin', '/bin']" "['/usr/sbin', '/bin', '${iproute2}/bin', '${pkgs.systemd}/bin']"
```
Comment (SuperSandro2000) [inline pkgs/tools/virtualization/cloud-init/default.nix:33]:
```suggestion
      --replace "['/usr/sbin', '/bin']" "['/usr/sbin', '/bin', '${iproute2}/bin', '${systemd}/bin']"
```

File: pkgs/tools/virtualization/cloud-init/default.nix [hunk 2]
```diff
@@ -1,8 +1,11 @@
 { lib
+, pkgs
```
Comment (SuperSandro2000) [inline pkgs/tools/virtualization/cloud-init/default.nix:2]:
```suggestion
```

File: pkgs/tools/virtualization/cloud-init/default.nix [hunk 3]
```diff
@@ -1,8 +1,11 @@
...
+, nixosTests
 , buildPythonApplication
 , cloud-utils
 , dmidecode
 , fetchFromGitHub
+, iproute2
 , openssh
 , python3
 , shadow
```
Comment (SuperSandro2000) [inline pkgs/tools/virtualization/cloud-init/default.nix:11]:
```suggestion
, shadow
, systemd
```

File: pkgs/tools/virtualization/cloud-init/default.nix [hunk 4]
```diff
@@ -25,6 +28,13 @@ python3.pkgs.buildPythonApplication rec {
   prePatch = ''
     substituteInPlace setup.py \
       --replace /lib/systemd $out/lib/systemd
+
+    substituteInPlace cloudinit/net/networkd.py \
+      --replace "['/usr/sbin', '/bin']" "['/usr/sbin', '/bin', '${iproute2}/bin', '${pkgs.systemd}/bin']"
+
+    substituteInPlace tests/unittests/test_net_activators.py \
+      --replace "['/usr/sbin', '/bin']" \
+        "['/usr/sbin', '/bin', '${iproute2}/bin', '${pkgs.systemd}/bin']"
```
Comment (SuperSandro2000) [inline pkgs/tools/virtualization/cloud-init/default.nix:37]:
```suggestion
      --replace "['/usr/sbin', '/bin']" "['/usr/sbin', '/bin', '${iproute2}/bin', '${systemd}/bin']"
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/tools/virtualization/cloud-init/default.nix (hunk 1 above)
Comment (illustris) [reply pkgs/tools/virtualization/cloud-init/default.nix:33]: Tried this, will not work. `systemd` refers to `python3Packages.systemd`. Is there a better way to do this?

Comment (SuperSandro2000) [reply pkgs/tools/virtualization/cloud-init/default.nix]: inherit pkgs.systemd in python-packages.

**Pull request merged**